include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+                           const Function<spacedim, typename VectorType::value_type> &function,
+                           VectorType                             &vec_result,
+                           const bool                             enforce_zero_boundary,
+                           const Quadrature<dim-1>                &q_boundary,
+                           const bool                             project_to_boundary_first)
+    {
+      switch (dof.get_fe().n_components())
+        {
+        case 1:
+          project_parallel_helper<1>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 2:
+          project_parallel_helper<2>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 3:
+          project_parallel_helper<3>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 4:
+          project_parallel_helper<4>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
+        }
+    }
+
+
+
+    template <int dim, typename VectorType>
+    void project_matrix_free (const Mapping<dim>            &mapping,
+                              const DoFHandler<dim>         &dof,
+                              const ConstraintMatrix        &constraints,
+                              const Quadrature<dim>         &quadrature,
+                              const Function<dim, typename VectorType::value_type> &function,
+                              VectorType                    &vec_result,
+                              const bool                    enforce_zero_boundary,
+                              const Quadrature<dim-1>       &q_boundary,
+                              const bool                    project_to_boundary_first)
+    {
+      // If we can, use the matrix-free implementation
+      bool use_matrix_free = true;
+      // enforce_zero_boundary and project_to_boundary_first
+      // are not yet supported
+      if (enforce_zero_boundary || project_to_boundary_first)
+        use_matrix_free = false;
+      else
+        {
+          // Find out if the FiniteElement is supported
+          // This is copied from matrix_free/shape_info.templates.h
+          // and matrix_free/matrix_free.templates.h
+          if (dof.get_fe().degree==0 || dof.get_fe().n_base_elements()!=1)
+            use_matrix_free = false;
+          else
+            {
+              const FiniteElement<dim> *fe_ptr = &dof.get_fe().base_element(0);
+              if (fe_ptr->n_components() != 1)
```

Comment:
No, this comes from `shape_info.templates.h` as well. In particular, for `FESystem<dim>(FE_Q<dim>,dim)` the base element is still `FE_Q` which has only one component.

If this were the number of components of the `FiniteElement`, the `project_parallel_q_*` tests would certainly fail.